### PR TITLE
[SYCL] Fix SYCLLowerWGLocalMemoryPass crash

### PR DESCRIPTION
**Pull request by mlychkov**
When call of __sycl_allocateLocalMemory is lowered and erased from parent
function block then iterator on this call instruction may be invalidated.
Modify iterating through __sycl_allocateLocalMemory function users.

Signed-off-by: Mikhail Lychkov <mikhail.lychkov@intel.com>

### CODE REVIEW

File: llvm/lib/SYCLLowerIR/LowerWGLocalMemory.cpp [hunk 1]
```diff
@@ -98,8 +98,8 @@ static bool allocaWGLocalMemory(Module &M) {
 
   assert(ALMFunc->isDeclaration() && "should have declaration only");
 
-  for (User *U : ALMFunc->users()) {
-    auto *CI = cast<CallInst>(U);
+  for (auto U = ALMFunc->user_begin(), UE = ALMFunc->user_end(); U != UE;) {
```
Comment (DenisBakhvalov) [inline llvm/lib/SYCLLowerIR/LowerWGLocalMemory.cpp:101]: I think a conventional way to do this is to employ a worklist of instructions and then iterate over it without worrying about invalidating iterators.

**Review by DenisBakhvalov — CHANGES_REQUESTED** [review]
Please add a test to prevent regressions in the future.

File: llvm/lib/SYCLLowerIR/LowerWGLocalMemory.cpp (hunk 1 above)
Comment (AlexeySachkov) [reply llvm/lib/SYCLLowerIR/LowerWGLocalMemory.cpp:101]: Another possible approach is to collect instructions which you need to remove to some container and remove them later in a separate loop:

https://github.com/intel/llvm/blob/691f8428ad87c27136e6aa2a01729b86d485c03e/llvm/tools/sycl-post-link/SpecConstants.cpp#L622-L626

**Comment by mlychkov** [comment]
> Please add a test to prevent regressions in the future.

It is present already. This is llvm/test/SYCLLowerIR/group_local_memory.ll test which is failed on windows with clang debug version of compiler.

**Review by DenisBakhvalov — APPROVED** [review]

**Review by AlexeySachkov — APPROVED** [review]

**Pull request merged**